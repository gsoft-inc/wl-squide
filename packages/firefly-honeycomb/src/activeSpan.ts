import type { Span } from "@opentelemetry/api";
import type { FetchRequestHookFunction } from "@opentelemetry/instrumentation-fetch";
import { isPlainObject } from "@squide/core";
import type { RuntimeLogger } from "@squide/firefly";
import { v4 as uuidv4 } from "uuid";
import { createTraceContextId } from "./createTraceContextId.ts";

declare global {
    interface Window { __SQUIDE_HONEYCOMB_ACTIVE_SPAN_STACK__: ActiveSpanStack }
}

const GlobalActiveSpanStackVariableName = "__SQUIDE_HONEYCOMB_ACTIVE_SPAN_STACK__";

export type ActiveSpanId = string;

export interface ActiveSpan {
    id: ActiveSpanId;
    name: string;
    instance: Span;
}

// Using a stack because we want a Last In First Out implementation for this.
// https://github.com/open-telemetry/opentelemetry-js/issues/5084
// https://github.com/open-telemetry/opentelemetry-js/issues/3558#issuecomment-1760680244
class ActiveSpanStack {
    readonly #stack: ActiveSpan[] = [];

    push(span: ActiveSpan) {
        this.#stack.push(span);
    }

    pop(span: ActiveSpan) {
        const head = this.#stack.pop();

        if (!head) {
            throw new Error("[honeycomb] Unexpected pop, the active span stack is empty.");
        }

        if (head.id !== span.id) {
            throw new Error(`[honeycomb] The active span is not the expected span. Expected to pop span with name and id "${span.name} / ${span.id}" but found "${head.name} / ${head.id}". Did you forget to end an active span?`);
        }

        return head;
    }

    peek() {
        if (this.#stack.length === 0) {
            return undefined;
        }

        return this.#stack[this.#stack.length - 1];
    }
}

export function registerActiveSpanStack() {
    if (window[GlobalActiveSpanStackVariableName]) {
        throw new Error(`[honeycomb] An ActiveSpanStack instance has already been registered to window.${GlobalActiveSpanStackVariableName}. Did you register the Honeycomb instrumentation twice?`);
    }

    window[GlobalActiveSpanStackVariableName] = new ActiveSpanStack();
}

function getActiveSpanStack() {
    return window[GlobalActiveSpanStackVariableName] as ActiveSpanStack;
}

function getActiveSpan() {
    const stack = getActiveSpanStack();

    if (stack) {
        return stack.peek();
    }
}

export function setActiveSpan(name: string, span: Span) {
    const activeSpan: ActiveSpan = {
        id: uuidv4(),
        name: name,
        instance: span
    };

    const stack = getActiveSpanStack();

    if (stack) {
        stack.push(activeSpan);
    }

    return activeSpan;
}

export function popActiveSpan(span: ActiveSpan) {
    const stack = getActiveSpanStack();

    if (stack) {
        stack.pop(span);
    }
}

let overrideFetchRequestSpanWithActiveSpanContextMock: FetchRequestHookFunction | undefined;

// This function should only be used by tests.
export function __setOverrideFetchRequestSpanWithActiveSpanContextMock(fct: FetchRequestHookFunction) {
    overrideFetchRequestSpanWithActiveSpanContextMock = fct;
}

// This function should only be used by tests.
export function __clearOverrideFetchRequestSpanWithActiveSpanContextMock() {
    overrideFetchRequestSpanWithActiveSpanContextMock = undefined;
}

export function createOverrideFetchRequestSpanWithActiveSpanContext(logger: RuntimeLogger) {
    if (overrideFetchRequestSpanWithActiveSpanContextMock) {
        return overrideFetchRequestSpanWithActiveSpanContextMock;
    }

    const fct: FetchRequestHookFunction = (span, request) => {
        const activeSpan = getActiveSpan();

        if (activeSpan) {
            const activeSpanContext = activeSpan.instance.spanContext();
            const requestSpanContext = span.spanContext();

            if (activeSpanContext) {
                logger.debug(
                    "[squide] Found a Honeycomb active context to apply to the following fetch request: \r\n",
                    "Request span context: ", requestSpanContext, "\r\n",
                    "Active span context: ", activeSpanContext, "\r\n",
                    "Request: ", request, "\r\n"
                );

                span.setAttribute("trace.trace_id", activeSpanContext.traceId);
                span.setAttribute("trace.parent_id", activeSpanContext.spanId);

                const traceParent = createTraceContextId(activeSpanContext.traceId, requestSpanContext.spanId, requestSpanContext.traceFlags);

                if (request instanceof Request) {
                    request.headers.set("traceparent", traceParent);
                } else if (isPlainObject(request.headers)) {
                    request.headers["traceparent"] = traceParent;
                }
            }
        }
    };

    return fct;
}
