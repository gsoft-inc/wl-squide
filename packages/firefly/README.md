# @squide/firefly

## Usage

View the [user's documentation](https://workleap.github.io/wl-squide/).

## ü§ù Contributing

View the [contributor's documentation](../../CONTRIBUTING.md).

## License

Copyright ¬© 2024, Workleap. This code is licensed under the Apache License, Version 2.0. You may obtain a copy of this license at https://github.com/workleap/workleap-license/blob/master/LICENSE.
