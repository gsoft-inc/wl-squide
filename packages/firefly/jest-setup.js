import "@testing-library/jest-dom";

import { TextDecoder, TextEncoder } from "node:util";

// https://github.com/mswjs/msw/issues/1796
global.TextEncoder = TextEncoder;
global.TextDecoder = TextDecoder;
