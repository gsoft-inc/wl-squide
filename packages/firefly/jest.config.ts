import type { Config } from "jest";
import { swcConfig } from "./swc.jest.ts";

const config: Config = {
    testRegex: "/tests/*/.*\\.test\\.(ts|tsx)$",
    testPathIgnorePatterns: ["/node_modules/", "/dist/"],
    testEnvironment: "jsdom",
    transform: {
        "^.+\\.(js|ts|tsx)$": ["@swc/jest", swcConfig as Record<string, unknown>]
    },
    transformIgnorePatterns: [
        "node_modules/(?!.pnpm|memoize|mimic-function)"
    ],
    globals: {
        __webpack_share_scopes__: {
            default: {}
        }
    },
    setupFilesAfterEnv: ["<rootDir>/jest-setup.js"],
    cacheDirectory: "./node_modules/.cache/jest",
    verbose: true
};

export default config;
