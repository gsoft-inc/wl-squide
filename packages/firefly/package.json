{
    "name": "@squide/firefly",
    "author": "Workleap",
    "version": "0.0.1",
    "description": "Helpers to facilitate the creation of a shell package with Squide firefly technology stack.",
    "license": "Apache-2.0",
    "repository": {
        "type": "git",
        "url": "git+https://github.com/gsoft-inc/wl-squide.git",
        "directory": "packages/firefly"
    },
    "publishConfig": {
        "access": "public",
        "provenance": true
    },
    "type": "module",
    "exports": {
        ".": {
            "import": "./dist/index.js",
            "types": "./dist/index.d.ts",
            "default": "./dist/index.js"
        }
    },
    "files": [
        "/dist",
        "CHANGELOG.md",
        "README.md"
    ],
    "scripts": {
        "dev": "tsup --config ./tsup.dev.ts",
        "build": "tsup --config ./tsup.build.ts"
    },
    "peerDependencies": {
        "msw": "*",
        "react": "*",
        "react-dom": "*",
        "react-error-boundary": "*",
        "react-router-dom": "*"
    },
    "peerDependenciesMeta": {
        "msw": {
            "optional": true
        }
    },
    "devDependencies": {
        "@swc/core": "1.3.95",
        "@swc/helpers": "0.5.3",
        "@swc/jest": "0.2.29",
        "@testing-library/react": "14.0.0",
        "@types/jest": "29.5.6",
        "@types/react": "18.2.33",
        "@types/react-dom": "18.2.14",
        "@workleap/eslint-plugin": "3.0.0",
        "@workleap/swc-configs": "2.1.2",
        "@workleap/tsup-configs": "3.0.1",
        "@workleap/typescript-configs": "3.0.2",
        "jest": "29.7.0",
        "jest-environment-jsdom": "29.7.0",
        "msw": "1.3.2",
        "react": "18.2.0",
        "react-dom": "18.2.0",
        "react-error-boundary": "4.0.11",
        "react-router-dom": "6.17.0",
        "ts-jest": "29.1.1",
        "tsup": "7.2.0",
        "typescript": "5.2.2"
    },
    "dependencies": {
        "@squide/msw": "workspace:*",
        "@squide/react-router": "workspace:*",
        "@squide/webpack-module-federation": "workspace:*"
    },
    "sideEffects": false,
    "engines": {
        "node": ">=18.0.0"
    }
}
