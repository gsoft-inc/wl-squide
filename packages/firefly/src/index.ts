export * from "@squide/core";
export * from "@squide/module-federation";
export * from "@squide/msw";
export * from "@squide/react-router";

export * from "./FireflyRuntime.tsx";

export * from "./AppRouter.tsx";
export * from "./AppRouterContext.ts";
export * from "./AppRouterReducer.ts";

export * from "./GlobalDataQueriesError.ts";
export * from "./useCanFetchProtectedData.ts";
export * from "./useCanFetchPublicData.ts";
export * from "./useCanRegisterDeferredRegistrations.ts";
export * from "./useCanUpdateDeferredRegistrations.ts";
export * from "./useDeferredRegistrations.ts";
export * from "./useFireflyNavigationItems.ts";
export * from "./useIsActiveRouteProtected.ts";
export * from "./useIsBootstrapping.ts";
export * from "./useProtectedDataQueries.ts";
export * from "./usePublicDataQueries.ts";
export * from "./useRegisterDeferredRegistrations.ts";
export * from "./useUpdateDeferredRegistrations.ts";

// TO DELETE
export * from "./LegacyAppRouter.tsx";

