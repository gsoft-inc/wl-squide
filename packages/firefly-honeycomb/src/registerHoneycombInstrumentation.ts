import { HoneycombWebSDK } from "@honeycombio/opentelemetry-web";
import type { Span } from "@opentelemetry/api";
import { getWebAutoInstrumentations, type InstrumentationConfigMap } from "@opentelemetry/auto-instrumentations-web";
import type { DocumentLoadInstrumentationConfig } from "@opentelemetry/instrumentation-document-load";
import type { FetchInstrumentationConfig } from "@opentelemetry/instrumentation-fetch";
import type { PropagateTraceHeaderCorsUrls, SpanProcessor } from "@opentelemetry/sdk-trace-web";
import {
    ApplicationBoostrappedEvent,
    ApplicationBootstrappingStartedEvent,
    LocalModuleDeferredRegistrationCompletedEvent,
    LocalModuleDeferredRegistrationFailedEvent,
    LocalModuleDeferredRegistrationStartedEvent,
    LocalModuleRegistrationCompletedEvent,
    LocalModuleRegistrationFailedEvent,
    LocalModuleRegistrationStartedEvent,
    ModulesReadyEvent,
    ModulesRegisteredEvent,
    MswReadyEvent,
    ProtectedDataFetchStartedEvent,
    ProtectedDataReadyEvent,
    PublicDataFetchStartedEvent,
    PublicDataReadyEvent,
    RemoteModuleDeferredRegistrationCompletedEvent,
    RemoteModuleDeferredRegistrationFailedEvent,
    RemoteModuleDeferredRegistrationStartedEvent,
    RemoteModuleRegistrationCompletedEvent,
    RemoteModuleRegistrationFailedEvent,
    RemoteModuleRegistrationStartedEvent,
    type FireflyRuntime,
    type LocalModuleDeferredRegistrationStartedEventPayload,
    type LocalModuleRegistrationStartedEventPayload,
    type RemoteModuleDeferredRegistrationStartedEventPayload,
    type RemoteModuleRegistrationStartedEventPayload
} from "@squide/firefly";
import { createApplyCustomAttributesOnFetchSpanFunction, registerActiveSpanStack, type ActiveSpan } from "./activeSpan.ts";
import { applyTransformers, type HoneycombSdkOptionsTransformer } from "./applyTransformers.ts";
import { globalAttributeSpanProcessor } from "./globalAttributes.ts";
import type { HoneycombSdkInstrumentations, HoneycombSdkOptions } from "./honeycombTypes.ts";
import { getTracer } from "./tracer.ts";
import { endActiveSpan, startActiveChildSpan, startChildSpan, startSpan, traceError } from "./utils.ts";

export type DefineFetchInstrumentationOptionsFunction = (defaultOptions: FetchInstrumentationConfig) => FetchInstrumentationConfig;
export type DefineDocumentLoadInstrumentationOptionsFunction = (defaultOptions: DocumentLoadInstrumentationConfig) => DocumentLoadInstrumentationConfig;

const defaultDefineFetchInstrumentationOptions: DefineFetchInstrumentationOptionsFunction = defaultOptions => {
    return defaultOptions;
};

const defaultDefineDocumentLoadInstrumentationOptions: DefineDocumentLoadInstrumentationOptionsFunction = defaultOptions => {
    return defaultOptions;
};

export interface RegisterHoneycombInstrumentationOptions {
    endpoint?: HoneycombSdkOptions["endpoint"];
    apiKey?: HoneycombSdkOptions["apiKey"];
    debug?: HoneycombSdkOptions["debug"];
    serviceName?: HoneycombSdkOptions["serviceName"];
    instrumentations?: HoneycombSdkInstrumentations;
    spanProcessors?: SpanProcessor[];
    fetchInstrumentation?: false | DefineFetchInstrumentationOptionsFunction;
    documentLoadInstrumentation?: false | DefineDocumentLoadInstrumentationOptionsFunction;
    transformers?: HoneycombSdkOptionsTransformer[];
}

export function registerHoneycombInstrumentation(runtime: FireflyRuntime, serviceName: NonNullable<HoneycombSdkOptions["serviceName"]>, apiServiceUrls: PropagateTraceHeaderCorsUrls, options: RegisterHoneycombInstrumentationOptions = {}) {
    const {
        endpoint,
        apiKey,
        debug: debugValue,
        instrumentations = [],
        spanProcessors = [],
        fetchInstrumentation = defaultDefineFetchInstrumentationOptions,
        documentLoadInstrumentation = defaultDefineDocumentLoadInstrumentationOptions,
        transformers = []
    } = options;

    // Defaults to the runtime mode.
    const debug = debugValue ?? runtime.mode === "development";

    if (!endpoint && !apiKey) {
        throw new Error("[honeycomb] Instrumentation must be initialized with either an \"endpoint\" or \"apiKey\" option.");
    }

    const instrumentationOptions = {
        ignoreNetworkEvents: true,
        propagateTraceHeaderCorsUrls: apiServiceUrls
    };

    const autoInstrumentations: InstrumentationConfigMap = {};

    if (fetchInstrumentation) {
        autoInstrumentations["@opentelemetry/instrumentation-fetch"] = fetchInstrumentation({
            ...instrumentationOptions,
            applyCustomAttributesOnSpan: createApplyCustomAttributesOnFetchSpanFunction(runtime.logger)
        });
    }

    if (documentLoadInstrumentation) {
        autoInstrumentations["@opentelemetry/instrumentation-document-load"] = documentLoadInstrumentation(instrumentationOptions);
    }

    const sdkOptions = {
        endpoint: endpoint,
        apiKey,
        debug,
        localVisualizations: debug,
        serviceName,
        instrumentations: [getWebAutoInstrumentations({ ...autoInstrumentations, ...instrumentations })],
        spanProcessors: [globalAttributeSpanProcessor, ...spanProcessors]
    } satisfies HoneycombSdkOptions;

    const transformedOptions = applyTransformers(sdkOptions, transformers);
    const instance = new HoneycombWebSDK(transformedOptions);

    instance.start();

    registerTrackingListeners(runtime);
    registerActiveSpanStack();
}

type DataFetchState = "none" | "fetching-data" | "public-data-ready" | "protected-data-ready" | "data-ready";

function reduceDataFetchEvents(
    runtime: FireflyRuntime,
    onDataFetchingStarted: () => void,
    onDataReady: () => void,
    onPublicDataFetchStarted: () => void,
    onPublicDataReady: () => void,
    onProtectedDataFetchStarted: () => void,
    onProtedtedDataReady: () => void
) {
    let dataFetchState: DataFetchState = "none";

    runtime.eventBus.addListener(PublicDataFetchStartedEvent, () => {
        if (dataFetchState === "none") {
            dataFetchState = "fetching-data";
            onDataFetchingStarted();
        }

        onPublicDataFetchStarted();
    });

    runtime.eventBus.addListener(PublicDataReadyEvent, () => {
        onPublicDataReady();

        if (dataFetchState === "fetching-data") {
            dataFetchState = "public-data-ready";
        } else if (dataFetchState === "protected-data-ready") {
            dataFetchState = "data-ready";
            onDataReady();
        }
    });

    runtime.eventBus.addListener(ProtectedDataFetchStartedEvent, () => {
        if (dataFetchState === "none") {
            dataFetchState = "fetching-data";
            onDataFetchingStarted();
        }

        onProtectedDataFetchStarted();
    });

    runtime.eventBus.addListener(ProtectedDataReadyEvent, () => {
        onProtedtedDataReady();

        if (dataFetchState === "fetching-data") {
            dataFetchState = "protected-data-ready";
        } else if (dataFetchState === "public-data-ready") {
            dataFetchState = "data-ready";
            onDataReady();
        }
    });
}

function registerTrackingListeners(runtime: FireflyRuntime) {
    let bootstrappingSpan: Span;
    let localModuleRegistrationSpan: Span;
    let localModuleDeferredRegistrationSpan: Span;
    let remoteModuleRegistrationSpan: Span;
    let remoteModuleDeferredRegistrationSpan: Span;
    let dataFetchSpan: ActiveSpan;

    runtime.eventBus.addListener(ApplicationBootstrappingStartedEvent, () => {
        bootstrappingSpan = startSpan((options, context) => getTracer().startSpan("squide-boostrapping", options, context));
    });

    runtime.eventBus.addListener(ApplicationBoostrappedEvent, () => {
        if (bootstrappingSpan) {
            bootstrappingSpan.end();
        }
    });

    runtime.eventBus.addListener(LocalModuleRegistrationStartedEvent, (payload: unknown) => {
        const attributes = {
            "squide.module_count": (payload as LocalModuleRegistrationStartedEventPayload).moduleCount
        };

        bootstrappingSpan.addEvent("local-module-registration-started", attributes);

        localModuleRegistrationSpan = startChildSpan(bootstrappingSpan, (options, context) => {
            return getTracer().startSpan("local-module-registration", { ...options, attributes }, context);
        });
    });

    runtime.eventBus.addListener(LocalModuleRegistrationCompletedEvent, () => {
        bootstrappingSpan.addEvent("local-module-registration-completed");

        if (localModuleRegistrationSpan) {
            localModuleRegistrationSpan.end();
        }
    });

    runtime.eventBus.addListener(LocalModuleRegistrationFailedEvent, (payload: unknown) => {
        if (localModuleRegistrationSpan) {
            traceError(localModuleRegistrationSpan, payload as Error);
        }
    });

    runtime.eventBus.addListener(LocalModuleDeferredRegistrationStartedEvent, (payload: unknown) => {
        const attributes = {
            "squide.registration_count": (payload as LocalModuleDeferredRegistrationStartedEventPayload).registrationCount
        };

        bootstrappingSpan.addEvent("local-module-deferred-registration-started", attributes);

        localModuleDeferredRegistrationSpan = startChildSpan(bootstrappingSpan, (options, context) => {
            return getTracer().startSpan("local-module-deferred-registration", { ...options, attributes }, context);
        });
    });

    runtime.eventBus.addListener(LocalModuleDeferredRegistrationCompletedEvent, () => {
        bootstrappingSpan.addEvent("local-module-deferred-registration-completed");

        if (localModuleDeferredRegistrationSpan) {
            localModuleDeferredRegistrationSpan.end();
        }
    });

    runtime.eventBus.addListener(LocalModuleDeferredRegistrationFailedEvent, (payload: unknown) => {
        if (localModuleDeferredRegistrationSpan) {
            traceError(localModuleRegistrationSpan, payload as Error);
        }
    });

    runtime.eventBus.addListener(RemoteModuleRegistrationStartedEvent, (payload: unknown) => {
        const attributes = {
            "squide.remote_count": (payload as RemoteModuleRegistrationStartedEventPayload).remoteCount
        };

        bootstrappingSpan.addEvent("remote-module-registration-started", attributes);

        remoteModuleRegistrationSpan = startChildSpan(bootstrappingSpan, (options, context) => {
            return getTracer().startSpan("remote-module-registration", { ...options, attributes }, context);
        });
    });

    runtime.eventBus.addListener(RemoteModuleRegistrationCompletedEvent, () => {
        bootstrappingSpan.addEvent("remote-module-registration-completed");

        if (remoteModuleRegistrationSpan) {
            remoteModuleRegistrationSpan.end();
        }
    });

    runtime.eventBus.addListener(RemoteModuleRegistrationFailedEvent, (payload: unknown) => {
        if (remoteModuleRegistrationSpan) {
            traceError(remoteModuleRegistrationSpan, payload as Error);
        }
    });

    runtime.eventBus.addListener(RemoteModuleDeferredRegistrationStartedEvent, (payload: unknown) => {
        const attributes = {
            "squide.registration_count": (payload as RemoteModuleDeferredRegistrationStartedEventPayload).registrationCount
        };

        bootstrappingSpan.addEvent("remote-module-deferred-registration-started", attributes);

        remoteModuleDeferredRegistrationSpan = startChildSpan(bootstrappingSpan, (options, context) => {
            return getTracer().startSpan("remote-module-deferred-registration", { ...options, attributes }, context);
        });
    });

    runtime.eventBus.addListener(RemoteModuleDeferredRegistrationCompletedEvent, () => {
        bootstrappingSpan.addEvent("remote-module-deferred-registration-completed");

        if (remoteModuleDeferredRegistrationSpan) {
            remoteModuleDeferredRegistrationSpan.end();
        }
    });

    runtime.eventBus.addListener(RemoteModuleDeferredRegistrationFailedEvent, (payload: unknown) => {
        if (remoteModuleDeferredRegistrationSpan) {
            traceError(remoteModuleDeferredRegistrationSpan, payload as Error);
        }
    });

    const handleFetchDataStarted = () => {
        dataFetchSpan = startActiveChildSpan(bootstrappingSpan, (options, context) => {
            const name = "data-fetch";
            const span = getTracer().startSpan(name, options, context);

            return {
                name,
                span
            };
        });
    };

    const handleDataReady = () => {
        if (dataFetchSpan) {
            endActiveSpan(dataFetchSpan);
        }
    };

    const handlePublicDataFetchStarted = () => {
        if (dataFetchSpan) {
            dataFetchSpan.instance.addEvent("public-data-fetch-started");
        }
    };

    const handlePublicDataReady = () => {
        if (dataFetchSpan) {
            dataFetchSpan.instance.addEvent("public-data-ready");
        }
    };

    const handleProtectedDataFetchStarted = () => {
        if (dataFetchSpan) {
            dataFetchSpan.instance.addEvent("protected-data-fetch-started");
        }
    };

    const handleProtectedDataReady = () => {
        if (dataFetchSpan) {
            dataFetchSpan.instance.addEvent("protected-data-ready");
        }
    };

    reduceDataFetchEvents(
        runtime,
        handleFetchDataStarted,
        handleDataReady,
        handlePublicDataFetchStarted,
        handlePublicDataReady,
        handleProtectedDataFetchStarted,
        handleProtectedDataReady
    );

    runtime.eventBus.addListener(ModulesRegisteredEvent, () => {
        bootstrappingSpan.addEvent("modules-registered");
    });

    runtime.eventBus.addListener(ModulesReadyEvent, () => {
        bootstrappingSpan.addEvent("modules-ready");
    });

    runtime.eventBus.addListener(MswReadyEvent, () => {
        bootstrappingSpan.addEvent("msw-ready");
    });
}
