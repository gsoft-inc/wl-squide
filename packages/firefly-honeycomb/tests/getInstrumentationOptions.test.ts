import type { FetchInstrumentationConfig } from "@opentelemetry/instrumentation-fetch";
import { FireflyRuntime } from "@squide/firefly";
import { __clearOverrideFetchRequestSpanWithActiveSpanContextMock, __setOverrideFetchRequestSpanWithActiveSpanContextMock } from "../src/activeSpan.ts";
import { getInstrumentationOptions } from "../src/registerHoneycombInstrumentation.ts";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function removeInstrumentationVersionsForSnapshot(options: any) {
    if (Array.isArray(options.instrumentations)) {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        options.instrumentations.forEach(x => {
            if (x["instrumentationVersion"]) {
                delete x["instrumentationVersion"];
            }

            if (x["version"]) {
                delete x["version"];
            }

            if (x["_logger"] && x["_logger"]["version"]) {
                delete x["_logger"]["version"];
            }

            if (x["_tracer"] && x["_tracer"]["version"]) {
                delete x["_tracer"]["version"];
            }
        });
    }

    return options;
}

test("when debug is true", () => {
    const runtime = new FireflyRuntime();

    const result = getInstrumentationOptions(runtime, {
        debug: true,
        apiKey: "123"
    });

    const cleanedResult = removeInstrumentationVersionsForSnapshot(result);

    expect(cleanedResult).toMatchSnapshot();
});

test("when debug is false", () => {
    const runtime = new FireflyRuntime();

    const result = getInstrumentationOptions(runtime, {
        debug: false,
        apiKey: "123"
    });

    const cleanedResult = removeInstrumentationVersionsForSnapshot(result);

    expect(cleanedResult).toMatchSnapshot();
});

test("when the runtime mode is \"development\"", () => {
    const runtime = new FireflyRuntime({
        mode: "development"
    });

    const result = getInstrumentationOptions(runtime, {
        apiKey: "123"
    });

    const cleanedResult = removeInstrumentationVersionsForSnapshot(result);

    expect(cleanedResult).toMatchSnapshot();
});

test("when the runtime mode is \"production\"", () => {
    const runtime = new FireflyRuntime({
        mode: "production"
    });

    const result = getInstrumentationOptions(runtime, {
        apiKey: "123"
    });

    const cleanedResult = removeInstrumentationVersionsForSnapshot(result);

    expect(cleanedResult).toMatchSnapshot();
});

describe("fetchInstrumentation", () => {
    test("when fetchInstrumentation is false, return false", () => {
        const runtime = new FireflyRuntime();

        const result = getInstrumentationOptions(runtime, {
            fetchInstrumentation: false,
            apiKey: "123"
        });

        expect(result.fetchInstrumentation).toBe(false);
    });

    test("when fetchInstrumentation is a function, call the function with the augmented options", () => {
        const runtime = new FireflyRuntime();

        const mock = jest.fn();

        const result = getInstrumentationOptions(runtime, {
            fetchInstrumentation: mock,
            apiKey: "123"
        });

        // Simulate calling the "registerHoneycombInstrumentation" function.
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        result.fetchInstrumentation({});

        expect(mock).toHaveBeenCalledTimes(1);
        expect(mock).toHaveBeenCalledWith(expect.objectContaining({
            requestHook: expect.any(Function)
        }));
    });

    test("when fetchInstrumentation is not provided, requestHook is the active span override function", () => {
        const runtime = new FireflyRuntime();

        const result = getInstrumentationOptions(runtime, {
            apiKey: "123"
        });

        expect(result.fetchInstrumentation).toBeDefined();

        // Simulate calling the "registerHoneycombInstrumentation" function.
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        const fetchOptions = result.fetchInstrumentation({}) as FetchInstrumentationConfig;

        expect(fetchOptions.requestHook).toBeDefined();
    });

    test("when the base honeycomb instrumentation library configure a default requestHook, merge the base function with the active span override function", () => {
        const runtime = new FireflyRuntime();

        const baseConfigMock = jest.fn();
        const activeSpanMock = jest.fn();

        __setOverrideFetchRequestSpanWithActiveSpanContextMock(activeSpanMock);

        const result = getInstrumentationOptions(runtime, {
            apiKey: "123"
        });

        // Simulate calling the "registerHoneycombInstrumentation" function.
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        const { requestHook } = result.fetchInstrumentation({
            requestHook: baseConfigMock
        });

        requestHook();

        expect(baseConfigMock).toHaveBeenCalledTimes(1);
        expect(activeSpanMock).toHaveBeenCalledTimes(1);

        __clearOverrideFetchRequestSpanWithActiveSpanContextMock();
    });
});
