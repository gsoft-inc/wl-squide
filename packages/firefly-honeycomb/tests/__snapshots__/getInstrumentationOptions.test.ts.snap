// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when debug is false 1`] = `
{
  "apiKey": "123",
  "debug": false,
  "fetchInstrumentation": [Function],
}
`;

exports[`when debug is true 1`] = `
{
  "apiKey": "123",
  "debug": true,
  "fetchInstrumentation": [Function],
}
`;

exports[`when the runtime mode is "development" 1`] = `
{
  "apiKey": "123",
  "debug": true,
  "fetchInstrumentation": [Function],
}
`;

exports[`when the runtime mode is "production" 1`] = `
{
  "apiKey": "123",
  "debug": false,
  "fetchInstrumentation": [Function],
}
`;
